# 作業目標:
    運用 Keras 模組建構CNN, 了解 CNN 的架構
    
    

# 作業重點
(1)嘗試比對 Dense 與 layers.Conv2D 架構NN 的差異

(2) 有沒有Pooling layer, 對於參數量的差異

注意: input_shape 請勿修改

In [ ]:
#導入相關模組

import keras
from keras import layers
from keras import models
from keras.models import Sequential
from keras.layers import Conv2D, Activation, MaxPooling2D, Flatten, Dense

#確認keras 版本
print(keras.__version__)

2.8.0


**只有全連接層的NN架構**

In [ ]:
model = Sequential([
    Dense(32, input_shape=(784,)),  #28x28=784
    Dense(100),
    Activation('relu'),
    Dense(10),
    Activation('softmax'),
])

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 32)                25120     
                                                                 
 dense_1 (Dense)             (None, 100)               3300      
                                                                 
 activation (Activation)     (None, 100)               0         
                                                                 
 dense_2 (Dense)             (None, 10)                1010      
                                                                 
 activation_1 (Activation)   (None, 10)                0         
                                                                 
Total params: 29,430
Trainable params: 29,430
Non-trainable params: 0
_________________________________________________________________


一共有29430個參數

# layers.Conv2D 模型, 用作比對

![CNN_Model.png](attachment:CNN_Model.png)



In [ ]:

#建立一個序列模型
model = models.Sequential()

#建立兩個卷績層, 32 個內核, 內核大小 3x3, 
#輸入影像大小 28x28x1
model.add(layers.Conv2D(32, (3, 3), input_shape=(28, 28, 1)))

#建立第二個卷績層, 池化層, 
#請注意, 不需要再輸入 input_shape
model.add(layers.Conv2D(25, (3, 3)))  #32-28+1=25

#新增平坦層
model.add(Flatten())

#建立一個全連接層
model.add(Dense(units=100))
model.add(Activation('relu'))

#建立一個輸出層, 並採用softmax
model.add(Dense(units=10))
model.add(Activation('softmax'))

model.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 conv2d_3 (Conv2D)           (None, 24, 24, 25)        7225      
                                                                 
 flatten_1 (Flatten)         (None, 14400)             0         
                                                                 
 dense_5 (Dense)             (None, 100)               1440100   
                                                                 
 activation_4 (Activation)   (None, 100)               0         
                                                                 
 dense_6 (Dense)             (None, 10)                1010      
                                                                 
 activation_5 (Activation)   (None, 10)               

完全不加池化層，參數量達到1448655

In [ ]:
from keras.layers.pooling import MaxPool2D
#建立一個序列模型
model = models.Sequential()

#建立兩個卷績層, 32 個內核, 內核大小 3x3, 
#輸入影像大小 28x28x1
model.add(layers.Conv2D(32, (3, 3), input_shape=(28, 28, 1)))
model.add(MaxPool2D(2,2))

#建立第二個卷績層, 池化層, 
#請注意, 不需要再輸入 input_shape
model.add(layers.Conv2D(25, (3, 3)))  #32-28+1=25
model.add(MaxPool2D(2,2))

#新增平坦層
model.add(Flatten())

#建立一個全連接層
model.add(Dense(units=100))
model.add(Activation('relu'))

#建立一個輸出層, 並採用softmax
model.add(Dense(units=10))
model.add(Activation('softmax'))

model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_4 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 13, 13, 32)       0         
 )                                                               
                                                                 
 conv2d_5 (Conv2D)           (None, 11, 11, 25)        7225      
                                                                 
 max_pooling2d_1 (MaxPooling  (None, 5, 5, 25)         0         
 2D)                                                             
                                                                 
 flatten_2 (Flatten)         (None, 625)               0         
                                                                 
 dense_7 (Dense)             (None, 100)              

兩層卷積層後都加上一層池化層後，參數量下降至71555